src/openql_i.cc

```diff
@@ -257,19 +265,144 @@ void Kernel::gate(
     const std::string &name,
     const std::vector<size_t> &qubits,
     size_t duration,
-    double angle
+    double angle,
+    const std::vector<size_t> &bregs,
+    const std::string &condstring,
+    const std::vector<size_t> &condregs
 ) {
-    kernel->gate(name, {qubits.begin(), qubits.end()}, {}, duration, angle);
+    QL_DOUT(
+        "Python k.gate("
+        << name
+        << ", "
+        << ql::utils::Vec<size_t>(qubits.begin(), qubits.end())
+        << ", "
+        << duration
+        << ", "
+        << angle
+        << ", "
+        << ql::utils::Vec<size_t>(bregs.begin(), bregs.end())
+        << ", "
+        << condstring
+        << ", "
+        << ql::utils::Vec<size_t>(condregs.begin(), condregs.end())
+        << ")"
+    );
+    ql::cond_type_t condvalue;
+    if      (condstring == "COND_ALWAYS") condvalue = ql::cond_always;
+    else if (condstring == "COND_NEVER") condvalue = ql::cond_never;
+    else if (condstring == "COND") condvalue = ql::cond;
+    else if (condstring == "COND_NOT") condvalue = ql::cond_not;
+    else if (condstring == "COND_AND") condvalue = ql::cond_and;
+    else if (condstring == "COND_NAND") condvalue = ql::cond_nand;
+    else if (condstring == "COND_OR") condvalue = ql::cond_or;
+    else if (condstring == "COND_NOR") condvalue = ql::cond_nor;
+    else if (condstring == "COND_XOR") condvalue = ql::cond_xor;
+    else if (condstring == "COND_NXOR") condvalue = ql::cond_nxor;
+    else {
+        throw std::runtime_error("Error: Unknown condition " + condstring);
+    }
+    kernel->gate(
+        name,
+        {qubits.begin(), qubits.end()},
+        {},
+        duration,
+        angle,
+        {bregs.begin(), bregs.end()},
+        condvalue,
+        {condregs.begin(), condregs.end()}
+    );
 }
 
 void Kernel::gate(
     const std::string &name,
     const std::vector<size_t> &qubits,
     const CReg &destination
 ) {
+    QL_DOUT(
+        "Python k.gate("
+        << name
+        << ", "
+        << ql::utils::Vec<size_t>(qubits.begin(), qubits.end())
+        << ",  "
+        << (destination.creg)->id
+        << ") # (name,qubits,creg-destination)"
+    );
     kernel->gate(name, {qubits.begin(), qubits.end()}, {(destination.creg)->id} );
 }
 
+void Kernel::gate_preset_condition(
+    const std::string &condstring,
+    const std::vector<size_t> &condregs
+) {
+    QL_DOUT("Python k.gate_preset_condition("<<condstring<<", condregs)");
+    ql::cond_type_t condvalue;
+    if      (condstring == "COND_ALWAYS") condvalue = ql::cond_always;
```

Comment:
move repeated code block to function

Reply:
Yes, could have a try, although openql_i.cc is not a usual C++ implementation file, so other rules hold.
So must probably factor this code out and put it in src/kernel.*